### PR TITLE
FEAT:(KB) several categories per article, target self-service users

### PR DESCRIPTION
**Pull request by Rom1-B**
Allow multiple categories to be added to a knowledge base article.

Allow access to a knowledge base article by self-service users without it being in the FAQ.

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | !22497

### CODE REVIEW

File: inc/knowbaseitem_category.class.php [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+         echo '<form method="post" action="' . Toolbox::getItemTypeFormURL(__CLASS__) . '">';
+         echo "<div class='center'>";
+         echo "<table class=\"tab_cadre_fixe\">";
+         echo "<tr><th colspan=\"2\">";
+         echo  __('Add a category');
+         echo "</th><tr>";
+         echo "<tr class='tab_bg_2'><td>";
+         $rand = KnowbaseItemCategory::dropdown();
+         echo "</td><td>";
+         echo "<input type=\"submit\" name=\"add\" value=\""._sx('button', 'Add')."\" class=\"submit\">";
```
Comment (orthagh) [inline inc/knowbaseitem_category.class.php:181]: remove `class="submit"` and add `class="btn btn-primary"`

**Review by orthagh — APPROVED** [review]
Apart the button stylen ok for merge

File: inc/knowbaseitem.class.php [hunk 2]
```diff
@@ -573,12 +577,8 @@ static public function getVisibilityCriteria(bool $forceall = false): array {
          $where['OR'] = [
                'glpi_knowbaseitems.users_id'       => Session::getLoginUserID(),
                'glpi_knowbaseitems_users.users_id' => Session::getLoginUserID(),
+               'glpi_knowbaseitems.is_faq'         => 1,
```
Comment (cedric-anne) [inline inc/knowbaseitem.class.php:580]: Are you sure of this change ?

File: inc/knowbaseitem.class.php [hunk 3]
```diff
@@ -1883,14 +1906,6 @@ function rawSearchOptions() {
          'datatype'           => 'number'
       ];
 
-      $tab[] = [
```
Comment (cedric-anne) [inline inc/knowbaseitem.class.php:1886]: Maybe this search option should be replaced by another one.

File: inc/knowbaseitem.class.php [hunk 4]
```diff
@@ -2039,6 +2054,7 @@ public function getAnswer() {
    protected function getShowVisibilityDropdownParams() {
       $params = parent::getShowVisibilityDropdownParams();
       $params['right'] = ($this->getField('is_faq') ? 'faq' : 'knowbase');
+      $params['allusers'] = 1;
```
Comment (cedric-anne) [inline inc/knowbaseitem.class.php:2057]: `$params['right']` will be ignore if `$params['allusers'] = 1`.

File: inc/knowbaseitem_category.class.php [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+if (!defined('GLPI_ROOT')) {
+   die("Sorry. You can't access this file directly");
+}
+
+/**
+ *  Class KnowbaseItem_Category
+ *
+ *  @author Romain Biot <rbiot@teclib.com>
+ *
+ *  @since 9.5.6
```
Comment (cedric-anne) [inline inc/knowbaseitem_category.class.php:42]:
```suggestion
 *  @since 10.0.0
```

File: inc/knowbaseitem_category.class.php [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+ */
+class KnowbaseItem_Category extends CommonDBRelation {
```
Comment (cedric-anne) [inline inc/knowbaseitem_category.class.php:44]: I am afraid people will confuse `KnowbaseItemCategory` and `KnowbaseItem_Category`.
According to GLPI conventions, we should name this class `KnowbaseItem_KnowbaseItemCategory`. Seem a long class name, but I guess it is preferable. @trasher @orthagh ?

File: inc/knowbaseitem_category.class.php [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+   static public $checkItem_2_Rights  = self::HAVE_VIEW_RIGHT_ON_ITEM;
+
+   // From CommonDBTM
+   public $dohistory = true;
+
+   static $rightname = 'knowbase';
+
+   function canPurgeItem() {
+      return Session::haveRight(static::$rightname, UPDATE);
+   }
```
Comment (cedric-anne) [inline inc/knowbaseitem_category.class.php:60]: This should not be needed.

File: inc/knowbaseitem_category.class.php [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+            ]
+         ],
+         'WHERE'     => [],
+         'GROUPBY'   => [
+            $kbi_cat_table . '.id'
+         ]
+      ];
+      $where = [];
+
+      $items_id  = (int)$item->getField('id');
```
Comment (cedric-anne) [inline inc/knowbaseitem_category.class.php:90]: 1. `getField` method should not be used anymore (or only in a display context).
2. `id` field should already be casted as integer by the mysqli driver.
```suggestion
      $items_id  = $item->fields['id'];
```

File: install/migrations/update_9.5.x_to_10.0.0/knowbaseitem_category.php [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+   if ($DB->fieldExists('glpi_knowbaseitems', 'knowbaseitemcategories_id')) {
+      $iterator = $DB->request([
+         'SELECT' => ['id', 'knowbaseitemcategories_id'],
+         'FROM'   => 'glpi_knowbaseitems',
+         'WHERE'  => ['knowbaseitemcategories_id' => ['>', 0]]
+      ]);
+      if (count($iterator)) {
+         //migrate existing data
+         $migration->migrationOneTable('glpi_knowbaseitems_categories');
```
Comment (cedric-anne) [inline install/migrations/update_9.5.x_to_10.0.0/knowbaseitem_category.php:60]: This seems useless.
```suggestion
```

File: install/migrations/update_9.5.x_to_10.0.0/knowbaseitem_category.php [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+      `id` int NOT NULL AUTO_INCREMENT,
+      `knowbaseitems_id` int NOT NULL DEFAULT '0',
+      `knowbaseitemcategories_id` int NOT NULL DEFAULT '0',
+      PRIMARY KEY (`id`),
+      KEY `knowbaseitems_id` (`knowbaseitems_id`),
+      KEY `knowbaseitemcategories_id` (`knowbaseitemcategories_id`)
+      ) ENGINE = InnoDB ROW_FORMAT = DYNAMIC DEFAULT CHARSET = {$default_charset} COLLATE = {$default_collation};";
+   $DB->queryOrDie($query, "add table glpi_knowbaseitems_categories");
+
+   if ($DB->fieldExists('glpi_knowbaseitems', 'knowbaseitemcategories_id')) {
```
Comment (cedric-anne) [inline install/migrations/update_9.5.x_to_10.0.0/knowbaseitem_category.php:52]: `glpi_knowbaseitems_categories` table creation and `glpi_knowbaseitems.knowbaseitemcategories_id` field migration should be done in two separated blocks. Indeed, if `glpi_knowbaseitems.knowbaseitemcategories_id` migration fails, for any reason, replaying the migration will never replay this part, and field will not be dropped, which can cause issues.

```suggestion
}
if ($DB->fieldExists('glpi_knowbaseitems', 'knowbaseitemcategories_id')) {
```

File: inc/knowbaseitem_category.class.php (hunk 6 above)
Comment (orthagh) [reply inc/knowbaseitem_category.class.php:44]: agree

File: inc/knowbaseitem.class.php (hunk 2 above)
Comment (Rom1-B) [reply inc/knowbaseitem.class.php:580]: Currently, self-service users only have access to FAQ articles. The goal is to share with them, even if the article is not in the FAQ.

File: inc/knowbaseitem_category.class.php (hunk 7 above)
Comment (Rom1-B) [reply inc/knowbaseitem_category.class.php:60]: This is necessary to remove links between articles and categories, but maybe it's not the right way?

Comment (cedric-anne) [reply inc/knowbaseitem_category.class.php]: I do not really know. Rights in CommonDBRelation are a pain to understand and debug.

**Review by cedric-anne — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

**Review by trasher — APPROVED** [review]

**Pull request merged**